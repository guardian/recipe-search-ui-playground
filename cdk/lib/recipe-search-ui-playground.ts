import type { GuStackProps} from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import type { App} from 'aws-cdk-lib';
import { aws_ssm, CfnOutput } from 'aws-cdk-lib';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { Distribution, HttpVersion } from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { Effect, PolicyStatement, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Bucket } from 'aws-cdk-lib/aws-s3';

const app = "recipe-search-ui-playground";

const hostingDomain:Record<string,string> = {
  "PROD": "recipe-search.capi.gutools.co.uk",
  "TEST": "recipe-search.test.capi.dev-gutools.co.uk"
}

export class RecipeSearchUiPlayground extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const webAclId = aws_ssm.StringParameter.valueForStringParameter(this, `/account/services/waf/default-ruleset-arn`);

    const staticBucketNameParam = new GuStringParameter(this, "StaticBucketName", {
      fromSSM: true,
      default: `/account/services/static.serving.bucket`,
      description: "SSM parameter giving the name of a bucket which is to be used for static hosting"
    });

    const hostingBucket = Bucket.fromBucketName(this, "StaticBucket", staticBucketNameParam.valueAsString);


    //We can't create the cert here, because it must live in us-east-1 for Cloudfront to use it.
    const hostingCertArn = new GuStringParameter(this, "DistroCertArn", {
      fromSSM: true,
      default: `/${this.stage}/${this.stack}/${app}/GlobalCertArn`,
      description: `Cert to use for recipe search UI playground ${this.stage}. This must reside in us-east-1`,
    });
    const hostingCert = Certificate.fromCertificateArn(this, "SearchPlaygroundCert", hostingCertArn.valueAsString);

    const cloudFrontDistro = new Distribution(this, "SearchPlaygroundDist", {
      certificate: hostingCert,
      defaultRootObject: "index.html",
      domainNames: [hostingDomain[this.stage]!],
      enableLogging: false,
      errorResponses: [],
      httpVersion: HttpVersion.HTTP2_AND_3,
      webAclId,
      defaultBehavior: {
        origin: new S3Origin(hostingBucket, {
          originPath: `/${this.stage}/${app}`
        })
      }
    });

    hostingBucket.addToResourcePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      principals: [
        new ServicePrincipal("cloudfront.amazonaws.com")
      ],
      actions: ["s3:GetObject"],
      resources: [`arn:aws:s3:::${hostingBucket.bucketName}/${app}/*`],
      conditions: {
        "StringEquals": {
          "AWS::SourceArn": `arn:aws:cloudfront::${this.account}:distribution/${cloudFrontDistro.distributionId}`
        }
      }
    }));

    new CfnOutput(this, "DistroUrlOut", {
      value: cloudFrontDistro.distributionDomainName,
    });

  }
}
