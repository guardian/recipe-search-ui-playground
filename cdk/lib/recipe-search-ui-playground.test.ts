import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { RecipeSearchUiPlayground } from "./recipe-search-ui-playground";

describe("The RecipeSearchUiPlayground stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new RecipeSearchUiPlayground(app, "RecipeSearchUiPlayground", { stack: "content-api", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
