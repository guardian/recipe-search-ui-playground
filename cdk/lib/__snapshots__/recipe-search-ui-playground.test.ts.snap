// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RecipeSearchUiPlayground stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStringParameter",
      "GuStringParameter",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "DistroUrlOut": {
      "Value": {
        "Fn::GetAtt": [
          "SearchPlaygroundDistE0868C1C",
          "DomainName",
        ],
      },
    },
  },
  "Parameters": {
    "DistroCertArn": {
      "Default": "/TEST/content-api/recipe-search-ui-playground/GlobalCertArn",
      "Description": "Cert to use for recipe search UI playground TEST. This must reside in us-east-1",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueaccountserviceswafdefaultrulesetarnC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/account/services/waf/default-ruleset-arn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "StaticBucketName": {
      "Default": "/account/services/static.serving.bucket",
      "Description": "SSM parameter giving the name of a bucket which is to be used for static hosting",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SearchPlaygroundDistE0868C1C": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "recipe-search.test.capi.dev-gutools.co.uk",
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "RecipeSearchUiPlaygroundSearchPlaygroundDistOrigin1A047ED52",
            "ViewerProtocolPolicy": "allow-all",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2and3",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Ref": "StaticBucketName",
                    },
                    ".s3.",
                    {
                      "Ref": "AWS::Region",
                    },
                    ".",
                    {
                      "Ref": "AWS::URLSuffix",
                    },
                  ],
                ],
              },
              "Id": "RecipeSearchUiPlaygroundSearchPlaygroundDistOrigin1A047ED52",
              "OriginPath": "/TEST/recipe-search-ui-playground",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "SearchPlaygroundDistOrigin1S3Origin26C237DE",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "DistroCertArn",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
          "WebACLId": {
            "Ref": "SsmParameterValueaccountserviceswafdefaultrulesetarnC96584B6F00A464EAD1953AFF4B05118Parameter",
          },
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/recipe-search-ui-playground",
          },
          {
            "Key": "Stack",
            "Value": "content-api",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "SearchPlaygroundDistOrigin1S3Origin26C237DE": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for RecipeSearchUiPlaygroundSearchPlaygroundDistOrigin1A047ED52",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
}
`;
